spec.md

```diff
@@ -281,6 +281,27 @@ help intermediate gateways determine how to route the events.
 As defined by the term [Data](#data), CloudEvents MAY include domain-specific
 information about the occurrence. When present, this information will be
 encapsulated within the `data` attribute.
+The `dataref` attribute can be used to reference another location where this
+information is stored. Known as the "Claim Check Pattern", the `dataref`
+attribute can be used for different purposes:
+
+* If the information is too large to be included in the message, the `data`
+  attribute is not present, and the consumer can retrieve it based on the
+  `dataref` attribute.
+* If the consumer wants to verify that the information has not been tampered
+  with, it can retrieve it from a trusted source using the `dataref` attribute.
+* If the information MUST only be viewed by trusted consumers (e.g. personally
+  identifiable information), only a trusted consumer can retrieve it using the
+  `dataref` attribute and a pre-shared secret.
+
+Both the `data` and `dataref` attribute MAY exist at the same time. A middleware
+MAY drop the `data` attribute when the `dataref` attribute exists, it MAY add
+the `dataref` attribute and drop the `data` attribute, or it MAY add the `data`
+attribute by using the `dataref` attribute.
```

Comment:
Just to complete all options... is there ever a case where they can drop `dataref` ?

Reply:
If we allow a middleware to drop the `dataref`, a consumer can not verify anymore that the `data` hasn't been tampered with. Less significantly, another middleware downstream can not (easily) drop the `data` attribute, if it wants to shrink the message size.

I can't come up with a good use case for a middleware to drop it. Did you have one in mind?

Reply:
Nope - was just asking for completeness. Thanks

Reply:
We need to be clear here on whether the dataref includes information carried in data attribute. If the dataref refers to information in addition to what is carried in data, then the middleware should not drop the data attribute. There are cases that it will result in performance enhancement if the event producer sends some key data information in-band in data attribute and other "might-need" large size information out-of-band via dataref. Extra latency is a drawback of serverless technology, so optimizing latency is important. There might be other use cases too. We may need to consider such use cases and allow data to be carried in both in-band data attribute and out-of-band dataref attribute.

Reply:
> We need to be clear here on whether the dataref includes information carried in data attribute.

I thought it should already be clear from the previous line, but apparently it is not. The information (as the spec calls it) of `data` and the one retrieved via `dataref` MUST be exactly the same.

I'll try to rephrase it.

> If the dataref refers to information in addition to what is carried in data, then the middleware should not drop the data attribute.

It must not.

> There are cases that it will result in performance enhancement if the event producer sends some key data information in-band in data attribute and other "might-need" large size information out-of-band via dataref.

Yes, this is a completely valid use case in itself, but not what `dataref` is. `dataref` is a reference to (what would be in) `data`.
If one wants to put some information within `data`, and some outside of it, that is completely fine - the references can be contained within `data`.
The point of `dataref` is that it allows a middleware (including a SDK) to switch between in-band and out-of-band without affecting consumers or producers.